control-plane/p4RuntimeSerializer.cpp

```diff
@@ -1316,6 +1367,15 @@ class P4RuntimeEntriesConverter {
             return stringRepr(k->to<IR::Constant>(), keyWidth);
         } else if (k->is<IR::BoolLiteral>()) {
             return stringRepr(k->to<IR::BoolLiteral>(), keyWidth);
+        } else if (k->type->is<IR::Type_Error>() ||
```

Comment:
Does P4Runtime even support keys of these types?

Reply:
P4Runtime v1.0 does not support keys of types other than the ones shown at this link: https://s3-us-west-2.amazonaws.com/p4runtime/docs/master/P4Runtime-Spec.html#sec-known-limitations-of-p4runtime-v100

and those same type restrictions exist for value_set match fields, action parameters, and fields of headers with an annotation `controller_header`, described elsewhere in the P4Runtime spec by searching for that string.

Reply:
PSA 1.1 does, but not p4runtime 1.0.  p4runtime does specify how to encode error and enum which is using human-readable string.  The code is being add for use with new PSA 1.1 P4 programs.

Reply:
This has nothing to do with PSA. This code should work for any architecture.

Reply:
P4Runtime is not the _only_ possible control plane API for P4_16 programs.  I am not aware of one, but perhaps one exists that allows table search key fields of type `error`.  It seems at least worth considering whether the part of p4c that determines which table search key fields are allowed, and which are not, should not be "too close" to the front end code.  I do not have any recommendation for where that list of allowed types belongs in the compiler's passes.

Reply:
This code is P4Runtime specific so please reject programs that are not supported according to the P4Runtime spec. Valid P4 programs that fail because they are not supported by P4Runtime can be listed in `P4RUNTIME_EXCLUDE` (https://github.com/p4lang/p4c/blob/master/backends/p4test/CMakeLists.txt#L60).

Reply:
Ok, will do.